src/docs/05-go-client/19-workflow-non-deterministic-error.md

```diff
@@ -0,0 +1,178 @@
+| 23 | TimerFired            | fired after 1 hour
+| 24 | DecisionTaskScheduled | triggered by TimerFired
+| 25 | DecisionTaskStarted   |
+| 26 | DecisionTaskCompleted |
+| 27 | WorkflowCompleted     | completed by decision
+
+As your may observe that this stack has strict orders. The whole point of the table above is that if the code you write involves some orchestration by Cadence, either your worker or Cadence server, they produce decision tasks. When your workflow gets replayed, it will strive to reconstruct this stack. Therefore, code changes to your workflow needs to make sure that they do not mess up with these decision tasks, which trigger non-deterministic errors. Then let's explore different types of non-detemistic errors and their root causes.
+
+## Categories of non-deterministic errors
+Programmatically, Cadence surfaces 4 categories of non-deterministic errors. With understanding of decision tasks in the previous section and combining the error messages, you should be able to pinpoint what code changes may yield to non-deterministic errors.
+
+### 1. Missing decisions
+```go
+fmt.Errorf("nondeterministic workflow: missing replay decision for %s", util.HistoryEventToString(e))
+```
+For source code click [here](https://github.com/uber-go/cadence-client/blob/e5081b085b0333bac23f198e57959681e0aee987/internal/internal_task_handlers.go#L1206)
+
+This means after replay code, the decision is scheduled less than history events. Using the previous history as an example, when the workflow is waiting at the one hour timer(event ID 22), if we delete the line of :
+```go
+workflow.Sleep(time.Hour)
+```
+and restart worker, then it will run into this error. Because in the history, the workflow has a timer event that is supposed to fire in one hour. However, during replay, there is no logic to schedule that timer.
+
+### 2. Extra decisions
+```go
+fmt.Errorf("nondeterministic workflow: extra replay decision for %s", util.DecisionToString(d))
+```
+For source code click [here](https://github.com/uber-go/cadence-client/blob/e5081b085b0333bac23f198e57959681e0aee987/internal/internal_task_handlers.go#L1210)
+
+This is basically the opposite of the previous case, which means that during replay, Cadence generates more decisions than those in history events. Using the previous history as an example, when the workflow is waiting at the one hour timer(event ID 22), if we change the line of:
+```go
+err = workflow.ExecuteActivity(ctx, activityB, a).Get(ctx, nil)
+```
+to
+```go
+fb := workflow.ExecuteActivity(ctx, activityB, a)
+fc := workflow.ExecuteActivity(ctx, activityC, a)
+err = fb.Get(ctx,nil)
+if err != nil {
+	return err
+}
+err = fc.Get(ctx,nil)
+if err != nil {
+	return err
+}
+```
+And restart worker, then it will run into this error. Because in the history, the workflow has scheduled only activityB after the one minute timer, however, during replay, there are two activities scheduled in a decision( in parallel).
```

Comment:
```suggestion
And restart worker, then it will run into this error. Because in the history, the workflow has scheduled only activityB after the one minute timer, however, during replay, there are two activities scheduled in a decision (in parallel).
```